src/OpenTelemetry.Instrumentation.EventCounters/EventCountersMetrics.cs

```diff
@@ -0,0 +1,137 @@
+// <copyright file="EventCountersMetrics.cs" company="OpenTelemetry Authors">
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// </copyright>
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Diagnostics.Metrics;
+using System.Diagnostics.Tracing;
+
+namespace OpenTelemetry.Instrumentation.EventCounters;
+
+/// <summary>
+/// .NET EventCounters Instrumentation.
+/// </summary>
+internal sealed class EventCountersMetrics : EventListener
+{
+    internal static readonly Meter MeterInstance = new(typeof(EventCountersMetrics).Assembly.GetName().Name, typeof(EventCountersMetrics).Assembly.GetName().Version.ToString());
+
+    private readonly EventCountersInstrumentationOptions options;
+    private readonly ConcurrentQueue<EventSource> preInitEventSources = new();
+    private readonly ConcurrentDictionary<(string, string), Instrument> instruments = new();
+    private readonly ConcurrentDictionary<(string, string), double> values = new();
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="EventCountersMetrics"/> class.
+    /// </summary>
+    /// <param name="options">The options to define the metrics.</param>
+    public EventCountersMetrics(EventCountersInstrumentationOptions options)
+    {
+        this.options = options;
+
+        while (this.preInitEventSources.TryDequeue(out EventSource eventSource))
+        {
+            if (this.options.ShouldListenToSource(eventSource.Name))
+            {
+                this.EnableEvents(eventSource, EventLevel.LogAlways, EventKeywords.None, GetEnableEventsArguments(this.options));
+            }
+        }
+    }
+
+    /// <inheritdoc />
+    protected override void OnEventSourceCreated(EventSource source)
+    {
+        if (this.options == null)
+        {
+            this.preInitEventSources.Enqueue(source);
+        }
+        else if (this.options.ShouldListenToSource(source.Name))
+        {
+            this.EnableEvents(source, EventLevel.LogAlways, EventKeywords.None, GetEnableEventsArguments(this.options));
+        }
+    }
+
+    /// <inheritdoc />
+    protected override void OnEventWritten(EventWrittenEventArgs eventData)
+    {
+        if (this.options == null)
```

Comment:
Is this check required?

Reply:
Ya, just dropping an event if the constructor hasn't finished setting the `options` field. I'll add comments as this is an obscure situation

Reply:
We never enable events until the options are set, right? If options is `null` we just enqueue it and wait till the ctor has set options before proceeding to enable events for the enqueued EventSources.

Reply:
Ya this is to avoid creating instruments for event sources we don't care about.